components/service/glean/src/test/java/mozilla/components/service/glean/ping/MetricsPingTest.kt

```diff
@@ -0,0 +1,57 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.glean.ping
+
+import android.content.Context
+import org.junit.Test
+import org.junit.Assert.assertFalse
+import org.junit.Assert.assertTrue
+import org.junit.runner.RunWith
+import org.robolectric.RobolectricTestRunner
+import org.robolectric.RuntimeEnvironment
+import java.util.concurrent.TimeUnit
+
+@RunWith(RobolectricTestRunner::class)
+class MetricsPingTest {
+    // NOTE: Using a "real" ApplicationContext here so that it will have
+    // a working SharedPreferences implementation
+    var applicationContext: Context = RuntimeEnvironment.application.applicationContext
+
+    @Test
+    fun `metrics ping has24HoursPassed() returns true if no value in shared prefs`() {
+        val metricsPing = MetricsPing(applicationContext)
+
+        // Clear the sharedPreferences on the MetricsPing so there is no last ping timestamp
+        metricsPing.metricsPreferences?.edit().clear().apply()
+
+        assertTrue(metricsPing.has24HoursPassed())
+    }
+
+    @Test
+    fun `metrics ping has24HoursPassed() returns false if 24 hours hasn't elapsed`() {
+        val metricsPing = MetricsPing(applicationContext)
+
+        // Clear the sharedPreferences on the MetricsPing so there is no last ping timestamp
+        metricsPing.metricsPreferences?.edit().clear().apply()
+
+        // Store a ping timestamp to compare to
+        metricsPing.updatePingTimestamp()
```

Comment:
I think you need to control the value you're writing here, otherwise you might end up with a flaky test. Moreover, since writing is done off-the main thread, I'd also make sure the expected value is written before doing the check below. Another way to do this more reliably would be to mock the preferences as [done here](https://github.com/mozilla-mobile/android-components/blob/74f0ae6e7ab84f18222518aadb445c51ea0af153/components/service/glean/src/test/java/mozilla/components/service/glean/storages/GenericScalarStorageEngineTest.kt#L152-L158), so that you're sure the value you want sits in there, avoiding potentially intermittent tests in the future.

Reply:
According to the documentation for [.apply()](https://developer.android.com/reference/android/content/SharedPreferences.Editor#apply()) the in memory SharedPreferences are updated immediately, it's just the write to disk that is async so there should be no way for the following read to fail since the in memory preferences are updated synchronously.